xnmt/decoder.py

```diff
@@ -113,10 +118,16 @@ def calc_loss(self, context, ref_action):
     scores = self.get_scores(context)
     # single mode
     if not xnmt.batcher.is_batched(ref_action):
-      return dy.pickneglogsoftmax(scores, ref_action)
-    # minibatch mode
+      log_loss = dy.pickneglogsoftmax(scores, ref_action)
+
+    else: # minibatch mode
+      log_loss = dy.pickneglogsoftmax_batch(scores, ref_action)
+
+    if self.label_smoothing == 0.0:
+      return log_loss
     else:
-      return dy.pickneglogsoftmax_batch(scores, ref_action)
+      ce_term = (1 - self.label_smoothing) * (dy.exp(-log_loss))
```

Comment:
Take another look at this and make sure it matches section 7 of https://arxiv.org/pdf/1512.00567.pdf